### PR TITLE
Warn users when using Campaign URL Builder for Google Play campaigns

### PR DESCRIPTION
**Pull request by Lucretiel**

### CODE REVIEW

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 1]
```diff
@@ -169,44 +208,71 @@ export default class CampaignUrl extends React.Component {
...
-          {this.renderActionsButton()}
+          {problematicElement}
+          {this.renderActionsButton(problematicElement ? true : false)}
           <div ref="url" className="u-visuallyHidden">{url}</div>
         </div>
       </div>
     );
   }
 
+  // NOTE(nathanwest): In the renderActionsButton, below, the IconButton
```
Comment (diminishedprime) [inline src/javascript/campaign-url-builder/components/campaign-url.js]: Not sure if this note is needed. I think it's safe to assume readers of this code are familiar React.

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 2]
```diff
@@ -71,6 +97,16 @@ export default class CampaignUrl extends React.Component {
     }
   }
 
+  confirmProblematic = () => {
+    gaSendEvent({
+      category: "CampaignUrl",
```
Comment (diminishedprime) [inline src/javascript/campaign-url-builder/components/campaign-url.js]: Single quotes here.

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 3]
```diff
@@ -169,44 +208,71 @@ export default class CampaignUrl extends React.Component {
...
   /**
    * Renders the Copy to Clipboard and Shorten URL buttons.
-   * @return {Object}
+   * @param {boolean} isProblematic If this parameter is true, the buttons
+   * are hidden behind an "I know what I'm doing" warning button.
+   * @return {JSX Literal}
    */
-  renderActionsButton() {
+  renderActionsButton(isProblematic) {
     return supports.copyToClipboard() ? (
```
Comment (diminishedprime) [inline src/javascript/campaign-url-builder/components/campaign-url.js]: This ternary is hard to parse, could you pull out the separate paths into named functions?

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ * Url Builder (https://goo.gl/qycyHd) instead of the GA devtools app.
+ *
+ * See b/69681865 for more details.
+ */
+
+import React from 'react';
+import {isRegExp} from 'lodash'
+
+
+// Create a function that checks if a url matches a regular expression.
```
Comment (diminishedprime) [inline src/javascript/campaign-url-builder/components/problematic.js]: Use JsDoc style comments for all function comments here and below.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 1 above)
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: True, but generally keys are only used to separate elements in arrays. This would be the first time I've ever come across their use outside of an array.

**Comment by Lucretiel** [comment]
Yeah, in retrospect I definitely agree about splitting it into multiple PRs. Think it's important enough to do in this case, or can we just say I've learned my lesson and won't do it again?

**Comment by diminishedprime** [comment]
lgtm

**Comment by Lucretiel** [comment]
Fixed linting issues for new / edited code. I can go and fix the unrelated stuff, too, since it looks fairly simple, if you like.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+const domainSuffix = suffix =>
+  domainMatches(`(?:[a-z0-9-]+.)*(?:${suffix})`);
+
+const googlePlayBuilderUrl = 'https://developers.google.com/analytics/devguides/collection/android/v4/campaigns#google-play-url-builder';
+const itunesStoreBuilderUrl = 'https://developers.google.com/analytics/devguides/collection/ios/v3/campaigns#url-builder';
+
+// Many fields in the badlist require a function with the signature
+// "url => thing". Use literal to make a function that just returns a
+// literal value
+const literal = thing => url => thing;
```
Comment (philipwalton) [inline src/javascript/campaign-url-builder/components/problematic.js]: This seems unnecessarily complex to me. Why not just use a `switch` statement?

While a `switch` statement would definitely be less clever, I think it would be clearer and not depend on a closure to do a pretty simple thing.

File: src/javascript/campaign-url-builder/index.js [hunk 6]
```diff
@@ -16,37 +16,14 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
 import {connect, Provider} from 'react-redux';
-import {bindActionCreators} from 'redux';
```
Comment (philipwalton) [inline src/javascript/campaign-url-builder/index.js]: Again, what's the rationale for removing redux? It's used in the other demos/tools, so removing it here doesn't save anything on file size, but it does create architectural differences between them.

File: src/javascript/campaign-url-builder/params.js [hunk 7]
```diff
@@ -14,6 +14,10 @@
 
 
 import qs from 'querystring';
+
+// TODO(nathanwest): url isn't actually a dependency of this project.
+// it seems like url-parse is the better library, anyway, so switch
+// to that at some point.
 import url from 'url';
```
Comment (philipwalton) [inline src/javascript/campaign-url-builder/params.js]: Sure, `parse-url` seems fine to me.

You could also use [this module](https://github.com/philipwalton/dom-utils/blob/master/lib/parse-url.js), which is already being loaded since it's used in autotrack, and it's smaller and has pretty much the same API.

```js
import parseUrl from 'dom-utils/lib/parse-url';
```

File: test/campaign-url-builder/problematic.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+          const renderResult = renderProblematic(url);
+          expect(renderResult)
+            .to.be.an('object')
+            .that.deep.equals({element: null, eventLabel: null});
+        });
+      });
+    });
+
+    it('returns a React element and analytics alert for problematic URLs',
+      () => {
```
Comment (philipwalton) [inline test/campaign-url-builder/problematic.js]: If you're going to line-break here, this should be indented 4 spaces.

File: test/campaign-url-builder/problematic.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        [
+          {domain: 'play.google.com', event: 'Google Play Store'},
+          {domain: 'itunes.apple.com', event: 'iOS App Store'},
+          {domain: 'ga-dev-tools.appspot.com', event: 'GA Dev Tools'},
+        ].forEach(({domain, event}) => {
+          thoroughlyTestDomain(domain, url => {
+            const renderResult = renderProblematic(url);
+            expect(renderResult)
+              .to.be.an('object')
```
Comment (philipwalton) [inline test/campaign-url-builder/problematic.js]: 4 spaces.

File: src/css/components/campaign-url-result.css [hunk 10]
```diff
@@ -23,6 +23,18 @@ limitations under the License.
   padding: var(--gap) var(--gap) 0;
 }
 
+.CampaignUrlResult-problem {
+  background: #fff3cd
+}
+
+.CampaignUrlResult-alert-box {
```
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/css/components/campaign-url-result.css]: It's also probably worth considering whether it makes sense to have a one-off component here for something that is functionally similar to the warning message on this page: https://ga-dev-tools.appspot.com/embed-api/server-side-authorization/

File: .gitignore [hunk 11]
```diff
@@ -17,6 +17,7 @@ python_modules
 
 # Built files
 public/
+build/
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply .gitignore]: I'm not rightly sure. It appears that all that happens the embed API build copies stuff from `public/javascript/embed-api/components/` to `build/javascript/embed-api/components`, and this is the only reference to `build/` anywhere in the project that I can find after about 10 minutes of searching. To be honest, I'd be about 80-90% confident we cant just remove the relevant section from the gulpfile, since not even the embed-api templates make reference to it.

For reference, this is the only reference I could find to that directory: https://github.com/googleanalytics/ga-dev-tools/blob/19dcf7c750af8214e5a306fc0f8e2b28bef7bb40/gulpfile.babel.js#L293-L296

File: .eslintrc [hunk 12]
```diff
@@ -36,16 +36,15 @@
...
-    // Unset rules conflicting with babel
-    "generator-star-spacing": 0,
     // react
     "react/jsx-key": 2,
     "react/jsx-uses-react": 2,
     "react/jsx-uses-vars": 2,
     "react/prefer-es6-class": [2, "always"],
     // babel
-    "babel/generator-star-spacing": [2, "after"]
+    "generator-star-spacing": [2, "after"],
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply .eslintrc]: Well, the babel rule was deprecated, so I just removed it and replaced it with the equivalent eslint rule. 

Fixed

File: gulpfile.babel.js [hunk 13]
```diff
@@ -273,11 +264,16 @@ gulp.task('test', ['lint'], function() {
...
-    if (data.includes('Starting module')) done();
+gulp.task('serve', [], done => {
+  const devServer = spawn('dev_appserver.py', [
+    '.',
+    '--host', process.env.GA_TOOLS_HOST || 'localhost',
+    '--port', process.env.GA_TOOLS_PORT || '8080',
+  ]);
+  devServer.stderr.on('data', data => {
+    if (data.includes('Starting module')) {
+done();
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply gulpfile.babel.js]: Fixed

File: .gitignore (hunk 11 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply .gitignore]: Ahh, now I remember. The build directory is for compiled versions of the Embed API plugin examples, so we should probably keep them.

Just revert this change and if you ever get conflicts in the build directory just don't worry about them.

[the remark this answers is not in the document]
Comment (Lucretiel) [reply .gitignore]: To be clear, the embed api builds to `public/`, and the HTML templates serve from there. There's an extra step in gulp to copy the files over, but I can't actually see that the files in `build/` are ever used for anything.

File: src/css/components/campaign-url-result.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/css/components/campaign-url-result.css]: We could try to deduplicate, but embed-api doesn't use React, so there's no component to duplicate.

File: .gitignore (hunk 11 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply .gitignore]: https://stackoverflow.com/questions/35047260/embed-api-custom-components-uncaught-typeerror-cannot-read-property-viewselec

File: src/css/components/campaign-url-result.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/css/components/campaign-url-result.css]: It doesn't need to be using React to share a common CSS class.

File: .gitignore (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply .gitignore]: That answer seems problematic, for two reasons:

- We shouldn't be telling people to download and use webpack output for things, especially not webpack output that's essentially ephemeral and specific to this project and not a purpose-build javascript library. There should be absolutely no expectation that build output can be used as some sort of functional drop-in library in people's projects.
- The answer as written is internally inconsistent. The text says to grab files from `build/...`, but the cited script tag sources content from `public/`.

File: src/css/components/campaign-url-result.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/css/components/campaign-url-result.css]: Oh, I misunderstood. Will fix.

[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/css/components/campaign-url-result.css]: So, the alertBox class I wrote was designed to be consistent with the bubble that already existed. I can copy the colors over, or even refactor, but that alert box (on the embed page) wouldn't really fit stylistically in the enclosing bubble

File: src/javascript/analytics.js [hunk 14]
```diff
@@ -66,16 +67,14 @@ export const NULL_VALUE = '(not set)';
...
-  return (command, ...args) => {
-    for (let {name} of trackers) {
-      if (typeof command == 'function') {
-        window.ga(() => command(window.ga.getByName(name)));
-      } else {
-        window.ga(`${name}.${command}`, ...args);
-      }
+const createGaProxy = trackers => (command, ...args) => {
+  for (const {name} of trackers) {
+    if (isFunction(command)) {
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/analytics.js]: Ah. I'm used to using extension for webpack (or maybe babel, I don't recall) that counteracts that, and automatically trims lodash imports to include only the stuff you explicitly import. It actually produces *smaller* binaries than doing the specific path'd import. Fixed.

I prefer isFunction because it's doing more stuff than just `typeof === function` ((source)[https://github.com/lodash/lodash/blob/4.17.4/lodash.js#L11661]), which implies to me that there's some kind of javascript nonsense you have to account for in order for it to be "correct". As a rule, I generally prefer the lodash type checking utilities for that reason.

File: src/javascript/campaign-url-builder/components/campaign-url-builder.js [hunk 15]
```diff
@@ -53,8 +53,8 @@ export default class CampaignUrlBuilder extends React.Component {
     let {params, bareUrl} = extractParamsFromWebsiteUrl(target.value);
 
     this.bareUrl = bareUrl;
-    this.props.actions.updateWebsiteUrl(target.value);
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url-builder.js]: I didn't move away from *redux* actions, but moving away from `props.actions` allows for substantially more idiomatic use of `react-redux`'s `connect` method:

> If an object is passed, each function inside it is assumed to be a Redux action creator. An object with the same function names, but with every action creator wrapped into a dispatch call so they may be invoked directly, will be merged into the component’s props. [[source](https://github.com/reactjs/react-redux/blob/master/docs/api.md#arguments)]

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 16]
```diff
@@ -33,16 +36,50 @@ const INSTRUCTIONS_TEXT =
     'automatically generated for you here.';
 
 
+const gaSendEvent = ({category, action, label}) =>
+  gaAll('send', 'event', {
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: I agree with you that it probably isn't necessary *in general*, for all the reasons you stated, which is why I didn't put it in the analytics module.

However, in this particular file, sending an event with category/action/label is both a common pattern (used half a dozen times in the file) and the only pattern that's ever used in the file, so I figured it made sense to reduce the amount of noise involved.

Plus it was bothering me to see the substring "event" 4 times in every call.

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 17]
```diff
@@ -33,16 +36,50 @@ const INSTRUCTIONS_TEXT =
...
+    const {element, eventLabel} = renderProblematic(props.url);
+    if (!isNil(element)) {
+      gaSendEvent({
+        category: 'Campaign URL',
+        action: 'Problematic URL',
+        label: eventLabel,
+      });
+    }
 
-  state = {
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Because I added a constructor, and part of what the constructor does is call `renderProblematic` with the props and populate the state with the results of that function call. If I wanted to be *really* more correct about it, I'd use a reselect selector instead, single the element and eventLabel are "derrived data" and not authoritative.

File: src/css/components/campaign-url-result.css (hunk 10 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/css/components/campaign-url-result.css]: I also considered the alert classes in `src/css/components/alert.css`, but those have a lot of flexbox attributes and probably are not suitable to just drop in anywhere.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * This file manages "problematic" URLs. For instance, if users are
+ * creating links to Play Store URLs, they should user the Google Play
+ * Url Builder (https://goo.gl/qycyHd) instead of the GA devtools app.
+ *
+ * See b/69681865 for more details.
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: I wanted to have it here for people to reference if they know how. I sought explicitly permission from my lead that it's okay to put these links in Open Source projects before putting it here.

File: src/javascript/campaign-url-builder/index.js (hunk 6 above)
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/index.js]: I didn't remove redux, I just removed `bindActionCreators`. The `connect` call still calls it internally, I just did a very minor shift to the idiomatic version.

Basically, you can do this:

```
Component = connect(..., {
    "action1": actionCreator1,
    "action2": actionCreator2,
})(BaseComponent)
```

And it automatically wraps each action creator in `dispatch` and merges the results into the props.

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 19]
```diff
@@ -169,44 +218,88 @@ export default class CampaignUrl extends React.Component {
...
-          {this.renderActionsButton()}
+          {problematicElement}
+          {this.renderButtons(problematicElement ? true : false)}
           <div ref="url" className="u-visuallyHidden">{url}</div>
         </div>
       </div>
     );
   }
 
+  // NOTE(nathanwest): In the renderActionsButton, below, the IconButton
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Fixed

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 20]
```diff
@@ -33,16 +36,50 @@ const INSTRUCTIONS_TEXT =
...
+
 /**
  * A component that renders the generated campaign URL.
  */
 export default class CampaignUrl extends React.Component {
+  /**
+   * Constructor for a campain URL component.
+   *
+   * @param {Object} props the initial props sent to the Rendered
+   * Campaign URL
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Fixed (I think). Let me know if I missed any after the push.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 19 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Fixed

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 21]
```diff
@@ -169,44 +218,88 @@ export default class CampaignUrl extends React.Component {
...
+        >
+          {this.state.isShorteningUrl ?
+            'Shortening...' : 'Convert URL to Short Link'}
+        </IconButton>
+      )}
+    </div>;
+  }
+
+  /**
+   * Render the button controls, under the URL. This will either be the
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Fixed. Weirdly, I was using the correct style everywhere else. I'm pretty sure this was the only case.

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 22]
```diff
@@ -169,44 +218,88 @@ export default class CampaignUrl extends React.Component {
...
-              type="link"
-              disabled={this.state.isShorteningUrl}
-              onClick={this.shortenUrl}>
-              {this.state.isShorteningUrl ?
-                'Shortening...' : 'Convert URL to Short Link'}
-            </IconButton>
-          )}
-        </div>
+        {isProblematic && !this.state.problematicBypass ?
+          this.renderProblematicWarningButton() :
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: What? Why? You mean the two expressions on the ternary operator?

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 23]
```diff
@@ -224,27 +317,51 @@ export default class CampaignUrl extends React.Component {
...
       this.setState({
         shortUrl: null,
         showShortUrl: false,
         isShorteningUrl: false,
+        problematicBypass: false,
+      });
+
+      // Compute the renderProblematic, and dispatch a GA event if the
+      // state went from Not Problematic to Problematic
+      // TODO(nathanwest): deduplicate this and the constructor code
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Fixed

File: src/javascript/campaign-url-builder/components/campaign-url.js [hunk 24]
```diff
@@ -224,27 +317,51 @@ export default class CampaignUrl extends React.Component {
...
+      const {element, eventLabel} = renderProblematic(nextProps.url);
+      this.setState(prevState => {
+        if (isNil(prevState.problematicElement) && !isNil(element)) {
+          gaSendEvent({
+            category: 'Campaign URL',
+            action: 'Problematic URL',
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: I don't understand what you mean by action like– you mean, the tag "Problematic URL" should be a verb?

File: src/javascript/campaign-url-builder/components/problematic.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: I figured it was better to have it then not have it, when Googlers read this code. I confirmed with my lead that it's acceptable to have links like this in Open Source code.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+ */
+
+import React from 'react';
+import {isRegExp} from 'lodash';
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: I actually like to order by "generalness", that is, system libs -> general libs (like lodash) -> specific libs (like react) -> local libs. But funnily enough, these imports to match that style, either.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+
+
+/**
+ * Create a function that, given a RegExp instance, returns a function
+ * that tests a url against that regex.
+ *
+ * @param  {RegExp} regexp compiled regular expression object used in
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: Oh, that wasn't deliberate; I can only assume my editor snippet was formatting it that way. Fixed.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 27]
```diff
@@ -0,0 +1,162 @@
...
+ * a given URL matches that pattern.
+ *
+ * @param  {string|Regexp} pattern to use in the returned testing
+ * function. If the pattern is a string, it is compiled (case-insensitive)
+ * to a Regexp
+ * @return {Function} function that tests if a URL matches
+ * the given regexp
+ */
+const matches = pattern => isRegExp(pattern) ?
+  matchesRegexp(pattern) :
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: Again, why? All the code in this project uses ubiquitous 2-space indentation, even at the root level

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 28]
```diff
@@ -0,0 +1,162 @@
...
+ * @param {string} suffix Suffix of the domain to match. Can be a pattern,
+ * but all . characters are unconditionally escaped
+ *
+ * @return {Function} a function that returns true if the
+ * domain matches the given url as a suffix
+ */
+const domainSuffix = suffix =>
+  domainMatches(`(?:[a-z0-9-]+.)*(?:${suffix})`);
+
+const googlePlayBuilderUrl = 'https://developers.google.com/analytics/devguides/collection/android/v4/campaigns#google-play-url-builder';
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: The URL lines? I agree, but I couldn't think of a way to make them shorter. This seemed to me like a reasonable exception to the line length rule, rather than trying to break up the URLs into multiple lines.

File: src/javascript/campaign-url-builder/components/problematic.js [hunk 29]
```diff
@@ -0,0 +1,162 @@
...
+ * be prepended with http:// or https://.
+ *
+ * @param {string} domain the domain to match. Can be a regex pattern, but all
+ * . characters are unconditionally escaped
+ *
+ * @return {Function} a function that returns true if the
+ * domain matches the given url exactly
+ */
+const domainMatches = domain =>
+  matches('^(?:https?://)?(:?' + domain.replace('.', '\\.') + ')(?:$|[/?#])');
```
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: I'm fine with it in general, but I fixed it for `domainMatches`

File: src/javascript/campaign-url-builder/components/problematic.js (hunk 5 above)
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: Yeah, I went back and forth on this. I liked how this made the renderProblematic function more simple, and the BADLIST more explicit, but I can revert it

Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: Oh, I also liked how it removed a conditional in `renderProblematic`; as a rule I like to avoid conditionals, if possible. But still, fixed.

File: src/javascript/campaign-url-builder/components/problematic.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/problematic.js]: I can only assume that eslint has an exception to the line length rule when the whole line is assigning a very long string to a variable.

File: src/javascript/analytics.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/analytics.js]: In general I don't have a problem with using this lodash methods if we're already using other lodash methods.

For some context, that file is based on my [analyticsjs-boilerplate](https://github.com/philipwalton/analyticsjs-boilerplate) project, and I like to keep it as close to the source as possible, so when I make upgrades to the boilerplate it's easy to upgrade them here as well.

As for your other comments. I get that concern, but I think most of the fear of JavaScript I hear these days is overblown. For example, Lodash's `isFunction()` handles cases where it tests against a constructor, but that's literally never going to happen in the above code.

My stance is that as long as your code is well tested and your tests pass in the environments you support, there's no need to use additional caution just out of fear. (Demos & Tools is not well-tested, so that's a different story.)

File: src/javascript/campaign-url-builder/components/campaign-url-builder.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url-builder.js]: Cool. I don't think that was a thing when this code was written.

File: .gitignore (hunk 11 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply .gitignore]: Yeah– I'd link to the real source files at https://github.com/googleanalytics/ga-dev-tools/blob/master/src/javascript/embed-api/components/view-selector2.js, and indicate that it's just an example of how to use `gapi.analytics.createComponent` to create a Component.

Because it has a dependency on `javascript-api-utils`, it's probably for the best to let the user resolve that dependency using whatever their own build system is, and direct them to resources on webpack if they're unfamiliar with how to use that.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: It's up to you, but I think it's better the other way around.

> Plus it was bothering me to see the substring "event" 4 times in every call.

FWIW keeping consistent use of the same fields object name will usually gzip better. By introducing another variation, you might be save unminified bytes but not necessarily gzipped size.

And as a side note, there's so much code and/or examples on the internet related to GA events that only shows category, action, and label. I think it creates this false impression that that's all there is. In devrel I think we should be promoting (or at least not blocking) advanced usage of our APIs `</rant>`.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Gotcha.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Yeah, according to Google JS style all line continuations are indented 4 spaces.

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: I mean the dimension name is "Event Action" and "Problematic URL" doesn't sound like an action to me. I was also saying it's probably worth taking a look at the other event actions we collect since whatever you choose, it's going to be in your reports forever.

File: src/javascript/campaign-url-builder/components/problematic.js (hunk 25 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/problematic.js]: Random info unrelated to this: Google JS style for internal code enforces module requires to be ordered alphabetically by namespace.

It doesn't currently say anything about ES module import statements, but I'd be surprised if it didn't eventually enforce alphabetical ordering in some form (either by module path or local variable name or something like that).

File: src/javascript/campaign-url-builder/components/problematic.js (hunk 28 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/problematic.js]: I think the style guide says to just use string concatenation. I know it makes an exception for URLs in comments, but I don't think it makes an exception for URLs in strings (though I'm not 100% sure about that).

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (Lucretiel) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: I guess I've always considered multi-line ternaries to be a form of block construct, akin to a minimalist if-else, and would fall under the 2-space rule. I can change it, though.

File: src/javascript/campaign-url-builder/index.js (hunk 6 above)
Comment (philipwalton) [reply src/javascript/campaign-url-builder/index.js]: Ok cool. It's been a while since I've looked at the developments in redux.

File: test/campaign-url-builder/problematic.js (hunk 8 above)
Comment (Lucretiel) [reply test/campaign-url-builder/problematic.js]: Do multi-line function arguments not count as a block construct? I figured that that parenthesis, and the fact that it isn't really an expression continutation, made the block rule apply

File: src/javascript/campaign-url-builder/components/campaign-url.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (philipwalton) [reply src/javascript/campaign-url-builder/components/campaign-url.js]: Yeah, I hate how it looks. But it's specifically shown that way in the style guide.

File: test/campaign-url-builder/problematic.js (hunk 8 above)
Comment (philipwalton) [reply test/campaign-url-builder/problematic.js]: I think the only things that count as block constructs are object and array literals.
